### PR TITLE
[EXPORTER] General cleanup for is_shutdown_ flags in exporters.

### PR DESCRIPTION
**Pull request by marcalff**
This is a follow up on PR #2553, to simplify `bool is_shutdown_` in general.

## Changes

Please provide a brief description of the changes here.

* Replace every use of a spin lock + boolean by an `std::atomic<bool>`
* Replace a spin lock by a mutex, when serializing to a file in the ostream metric exporter
* Implemented the missing call to `sout_.flush()` in `OStreamMetricExporter::ForceFlush()`

For significant contributions please make sure you have completed the following items:

* [ ] `CHANGELOG.md` updated for non-trivial changes
* [ ] Unit tests have been added
* [ ] Changes in public API reviewed

### CODE REVIEW

**Review by lalitb — APPROVED** [review]
Thanks for the cleanup.

File: exporters/ostream/include/opentelemetry/exporters/ostream/metric_exporter.h [hunk 1]
```diff
@@ -72,8 +73,8 @@ class OStreamMetricExporter final : public opentelemetry::sdk::metrics::PushMetr
 
 private:
   std::ostream &sout_;
-  bool is_shutdown_ = false;
-  mutable opentelemetry::common::SpinLockMutex lock_;
+  std::atomic<bool> is_shutdown_{false};
+  std::mutex serialize_lock_;
```
Comment (ThomsonTan) [inline exporters/ostream/include/opentelemetry/exporters/ostream/metric_exporter.h:77]: Why not using SpinLockMutext here?

Comment (lalitb) [reply exporters/ostream/include/opentelemetry/exporters/ostream/metric_exporter.h:77]: We were unfortunately using the `lock_` not just for shutdown :)

Comment (marcalff) [reply exporters/ostream/include/opentelemetry/exporters/ostream/metric_exporter.h:77]: The SpinLockMutex was also used to serialize when writing to sout_.

When the disk is full, a write or flush can hold forever, and using spin locks can make threads pilling up in CPU loops, which is very bad.

A mutex is more appropriate.

**Pull request merged**